core/trino-parser/src/test/java/io/trino/sql/parser/TestSqlParserErrorHandling.java

```diff
@@ -99,9 +99,9 @@ private static Stream<Arguments> statements()
                 Arguments.of("SELECT grouping(a+2) FROM (VALUES (1)) AS t (a) GROUP BY a+2",
                         "line 1:18: mismatched input '+'. Expecting: ')', ',', '.'"),
                 Arguments.of("SELECT x() over (ROWS select) FROM t",
-                        "line 1:23: mismatched input 'select'. Expecting: 'BETWEEN', 'CURRENT', 'UNBOUNDED', <expression>"),
+                        "line 1:23: mismatched input 'select'. Expecting: ')', 'BETWEEN', 'CURRENT', 'GROUPS', 'MEASURES', 'ORDER', 'PARTITION', 'RANGE', 'ROWS', 'UNBOUNDED', <expression>"),
                 Arguments.of("SELECT X() OVER (ROWS UNBOUNDED) FROM T",
-                        "line 1:32: mismatched input ')'. Expecting: 'FOLLOWING', 'PRECEDING'"),
+                        "line 1:32: mismatched input ')'. Expecting: '%', '(', '*', '+', '-', '->', '.', '/', 'AND', 'AT', 'FOLLOWING', 'OR', 'OVER', 'PRECEDING', '[', '||', <predicate>, <string>"),
```

Comment:
This doesn't seem right. The error position is reported as the end of `UNBOUNDED`:

```
SELECT X() OVER (ROWS UNBOUNDED) FROM T
                               ^
```

The only valid tokens at that point are `FOLLOWING` and `PRECEDING`, so we shouldn't be reporting all of these extra tokens.

Reply:
Yes, I know. I’m trying to figure out why it’s doing that. There’s some other bug lurking in there.

Reply:
It turns out these are all valid tokens per the grammar. It's because `UNBOUNDED` is a non-reserved word, so it can take the place of an identifier. In particular, it's considering alternatives that include the following productions:

```
over
    : OVER ([...] | '(' windowSpecification ')')
    ;

windowSpecification
    : [...]
      windowFrame?
    ;

windowFrame
    : [...]
      frameExtent
    ;

frameExtent
    : [...]
    | frameType=ROWS start=frameBound
    | [...]
    ;

frameBound
    : [...]
    | expression boundType=(PRECEDING | FOLLOWING)  #boundedFrame
    ;
```

And `<expression>` can be an `<identifier>` followed by a number of things, including `.`, `[`, `OVER`, etc.

For instance, this goes through the parser successfully even though it's clearly wrong if we consider `UNBOUNDED` a special keyword:

```
SELECT X() OVER (ROWS UNBOUNDED -> 1 PRECEDING) FROM t
```

It's treating `UNBOUNDED -> 1` as a lambda expression, which is a valid (syntax-wise) expression in the "#boundedFrame" frame bound above.

Reply:
If I mark `UNBOUNDED` as a reserved word, then it produces the expected message:

```
line 1:32: mismatched input ')'. Expecting: 'FOLLOWING', 'PRECEDING'
```

Reply:
Is there a way to make the error handler treat `UNBOUNDED` as a reserved word without actually making it reserved? In my opinion, printing all alternatives for the case of UNBOUNDED being an identifier is unnecessary and confusing, even if those are all correct per grammar.

Reply:
Another issue I can spot here is that the error handler suggests `GROUPS`, `RANGE`, and `ROWS` at position 23:
```
Arguments.of("SELECT x() over (ROWS select) FROM t",
                        "line 1:23: mismatched input 'select'. Expecting: 'BETWEEN', 'CURRENT', 'UNBOUNDED', <expression>"),
                        "line 1:23: mismatched input 'select'. Expecting: ')', 'BETWEEN', 'CURRENT', 'GROUPS', 'MEASURES', 'ORDER', 'PARTITION', 'RANGE', 'ROWS', 'UNBOUNDED', <expression>"),
```
It is not correct, as position 23 is after the word `ROWS`. It seems that the reported error position is the innermost one, but the suggestions come from an outer rule.